### PR TITLE
Enable CMake builds of C++ library.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This is an initial sketch of a `CMakeLists.txt` build file for freud's C++ components.

## Motivation and Context
It might be advantageous to use CMake with a system like [scikit-build](https://scikit-build.readthedocs.io/en/latest/) to build freud, because the C++ library could be built separately from the Cython modules. That would offer substantial performance improvements for build times. Furthermore, having a proper CMake build system allows for IDE integrations to analyze the library and make suggestions while developing the C++ code. That ease of development is the primary reason I decided to start this PR. Someday it may be worth exploring having a true C++ API for freud, but that's not yet something I am willing to consider stable.

## How Has This Been Tested?
*This will need to be thoroughly tested on multiple platforms before deployment to PyPI or conda-forge.*

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [ ] My code follows the code style of this project.
- [ ] I have updated the documentation (if relevant).
- [ ] I have added tests that cover my changes (if relevant).
- [ ] All new and existing tests passed.
- [ ] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [ ] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

### CODE REVIEW

**Comment by tommy-waltmann** [comment]
This new build configuration sounds a lot like what GSD implements. Before deciding whether to follow the GSD example or use `scikit-build`, it would be good to get some input from @joaander about the positives and negatives of GSD's build configuration.

Overall, I think this change will improve freud's build significantly. The C++ API is also a good idea, but it's something to leave for the future.

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                    "${CMAKE_SOURCE_DIR}/cpp/diffraction"
+                    "${CMAKE_SOURCE_DIR}/cpp/environment"
+                    "${CMAKE_SOURCE_DIR}/cpp/locality"
+                    "${CMAKE_SOURCE_DIR}/cpp/order"
+                    "${CMAKE_SOURCE_DIR}/cpp/parallel"
+                    "${CMAKE_SOURCE_DIR}/cpp/pmft"
+                    "${CMAKE_SOURCE_DIR}/cpp/util")
+
+# Define library
+add_library(freud SHARED
```
Comment (tommy-waltmann) [inline CMakeLists.txt:70]: I believe voro++'s source files need to be in this list also.

Comment (bdice) [reply CMakeLists.txt:70]: ~This is covered by `include_directories("extern")` and the corresponding `CMakeLists.txt` in that directory.~ Just kidding, you're probably right.

**Comment by vyasr** [comment]
Closing in favor of #668

**Pull request closed without merging**